plugins/aggregate/aggregate.cpp

```diff
@@ -0,0 +1,469 @@
+#include <vast/concept/parseable/to.hpp>
+#include <vast/concept/parseable/vast/time.hpp>
+#include <vast/detail/narrow.hpp>
+#include <vast/hash/hash_append.hpp>
+#include <vast/plugin.hpp>
+#include <vast/transform_step.hpp>
+
+#include <arrow/compute/api_aggregate.h>
+#include <arrow/compute/api_scalar.h>
+#include <arrow/compute/function.h>
+#include <arrow/table.h>
+#include <arrow/table_builder.h>
+
+#include <unordered_map>
+
+namespace vast::plugins::aggregate {
+
+/// The configuration of an aggregate transform step.
+struct configuration {
+  /// Duration window for grouping time values.
+  std::optional<duration> time_window = {};
+
+  /// List of fields to group by.
+  std::vector<std::string> group_by = {};
+
+  /// List of fields to sum.
+  std::vector<std::string> sum = {};
+
+  /// List of fields to take the minimum of.
+  std::vector<std::string> min = {};
+
+  /// List of fields to take the maximum of.
+  std::vector<std::string> max = {};
+
+  /// List of fields to take the maximum of.
+  std::vector<std::string> mean = {};
+
+  /// List of fields to take the disjunction of.
+  std::vector<std::string> any = {};
+
+  /// List of fields to take the conjunction of.
+  std::vector<std::string> all = {};
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, configuration& x) {
+    return f(x.time_window, x.group_by, x.sum, x.min, x.max, x.mean, x.any,
+             x.all);
+  }
+
+  static inline const record_type& layout() noexcept {
+    static auto result = record_type{
+      {"time-window", duration_type{}},  {"group-by", list_type{string_type{}}},
+      {"sum", list_type{string_type{}}}, {"min", list_type{string_type{}}},
+      {"max", list_type{string_type{}}}, {"mean", list_type{string_type{}}},
+      {"any", list_type{string_type{}}}, {"all", list_type{string_type{}}},
+    };
+    return result;
+  }
+};
+
+enum class action {
+  drop,
+  group_by_time_window,
+  group_by,
+  sum,
+  min,
+  max,
+  mean,
+  any,
+  all,
+};
+
+struct bucket_key : arrow::ScalarVector {
+  using arrow::ScalarVector::ScalarVector;
+
+  friend bool
+  operator==(const bucket_key& lhs, const bucket_key& rhs) noexcept {
+    return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(),
+                      [](const std::shared_ptr<arrow::Scalar>& x,
+                         const std::shared_ptr<arrow::Scalar>& y) noexcept {
+                        return x->Equals(y);
+                      });
+  }
+};
+
+} // namespace vast::plugins::aggregate
+
+namespace std {
+
+template <>
+struct std::hash<vast::plugins::aggregate::bucket_key> {
+  size_t
+  operator()(const vast::plugins::aggregate::bucket_key& x) const noexcept {
+    auto hasher = vast::xxh64{};
+    for (const auto& scalar : x) {
+      VAST_ASSERT(scalar);
+      vast::hash_append(hasher, scalar->hash());
+    }
+    return hasher.finish();
+  }
+};
+
+} // namespace std
+
+namespace vast::plugins::aggregate {
+
+struct resolved_columns {
```

Comment:
This struct has a really bad name. When @6yozo and I started prototyping this, it only held the resolved columns. Over time, we noticed that we have much more state that we want to cache per layout, and that is what this really is: our per-layout state.

Reply:
I renamed this into `aggregation`.